File: CSharp/Contests2022/ABC282/Lib/UnweightedGraph.cs
﻿using System;
using System.Collections.Generic;

// SPPs.Int.UnweightedGraph401 と同じ抽象グラフです。
// 頂点 id は整数 [0, n) とします。
namespace CoderLib8.Graphs.Specialized.Int
{
	[System.Diagnostics.DebuggerDisplay(@"VertexesCount = {VertexesCount}")]
	public abstract class UnweightedGraph
	{
		public int VertexesCount { get; }
		public abstract List<int> GetEdges(int v);
		protected UnweightedGraph(int vertexesCount) { VertexesCount = vertexesCount; }
	}

	public class ListUnweightedGraph : UnweightedGraph
	{
		protected readonly List<int>[] map;
		public List<int>[] AdjacencyList => map;
		public override List<int> GetEdges(int v) => map[v];

		public ListUnweightedGraph(List<int>[] map) : base(map.Length) { this.map = map; }
		public ListUnweightedGraph(int vertexesCount) : base(vertexesCount)
		{
			map = Array.ConvertAll(new bool[vertexesCount], _ => new List<int>());
		}
		public ListUnweightedGraph(int vertexesCount, IEnumerable<(int from, int to)> edges, bool twoWay) : this(vertexesCount)
		{
			foreach (var (from, to) in edges) AddEdge(from, to, twoWay);
		}

		public void AddEdge(int from, int to, bool twoWay)
		{
			map[from].Add(to);
			if (twoWay) map[to].Add(from);
		}
	}

	public class UnweightedGrid : UnweightedGraph
	{
		protected readonly int h, w;
		public int Height => h;
		public int Width => w;
		public UnweightedGrid(int h, int w) : base(h * w) { this.h = h; this.w = w; }

		public int ToVertexId(int i, int j) => w * i + j;
		public (int i, int j) FromVertexId(int v) => (v / w, v % w);

		public static (int di, int dj)[] NextsDelta { get; } = new[] { (0, -1), (0, 1), (-1, 0), (1, 0) };
		public static (int di, int dj)[] NextsDelta8 { get; } = new[] { (0, -1), (0, 1), (-1, 0), (1, 0), (-1, -1), (-1, 1), (1, -1), (1, 1) };

		public override List<int> GetEdges(int v)
		{
			var (i, j) = (v / w, v % w);
			var l = new List<int>();
			foreach (var (di, dj) in NextsDelta)
			{
				var (ni, nj) = (i + di, j + dj);
				if (0 <= ni && ni < h && 0 <= nj && nj < w) l.Add(w * ni + nj);
			}
			return l;
		}
	}

	public class CharUnweightedGrid : UnweightedGrid
	{
		readonly char[][] s;
		readonly char wall;
		public char[][] Cells => s;
		public char[] this[int i] => s[i];
		public CharUnweightedGrid(char[][] s, char wall = '#') : base(s.Length, s[0].Length) { this.s = s; this.wall = wall; }
		public CharUnweightedGrid(string[] s, char wall = '#') : this(ToArrays(s), wall) { }

		public static char[][] ToArrays(string[] s) => Array.ConvertAll(s, l => l.ToCharArray());

		public (int i, int j) FindCell(char c)
		{
			for (int i = 0; i < h; ++i)
				for (int j = 0; j < w; ++j)
					if (s[i][j] == c) return (i, j);
			return (-1, -1);
		}

		public int FindVertexId(char c)
		{
			for (int i = 0; i < h; ++i)
				for (int j = 0; j < w; ++j)
					if (s[i][j] == c) return w * i + j;
			return -1;
		}

		public override List<int> GetEdges(int v)
		{
			var (i, j) = (v / w, v % w);
			var l = new List<int>();
			foreach (var (di, dj) in NextsDelta)
			{
				var (ni, nj) = (i + di, j + dj);
				if (0 <= ni && ni < h && 0 <= nj && nj < w && s[ni][nj] != wall) l.Add(w * ni + nj);
			}
			return l;
		}
	}
}

File: Python/coder-lib-ws/lib/graphs/spps/unweightedgraph.py
from abc import ABCMeta, abstractmethod
from collections import deque
from collections.abc import Iterable


class UnweightedGraph(metaclass=ABCMeta):

    def __init__(self, n: int):
        self.n = n

    # -> Iterable[int] と書くと、AtCoder で RE になります。
    @abstractmethod
    def get_edges(self, v: int) -> Iterable:
        pass

    def connectivity_by_dfs(self, sv: int, ev=-1):
        u = [False] * self.n
        u[sv] = True

        def dfs(v: int) -> bool:
            if v == ev:
                return True
            for nv in self.get_edges(v):
                if u[nv]:
                    continue
                u[nv] = True
                if dfs(nv):
                    return True
            return False

        dfs(sv)
        return u

    def shortest_by_bfs(self, sv: int, ev=-1):
        c = [1 << 30] * self.n
        q: deque[int] = deque()
        c[sv] = 0
        q.append(sv)

        while q:
            v = q.popleft()
            if v == ev:
                return c

            for nv in self.get_edges(v):
                if c[nv] != 1 << 30:
                    continue
                c[nv] = c[v] + 1
                q.append(nv)
        return c


class ListUnweightedGraph(UnweightedGraph):

    def __init__(self, n: int):
        super().__init__(n)
        self.map: list[list[int]] = [[] for _ in range(n)]

    def get_edges(self, v: int):
        return self.map[v]

    def add_edge(self, u: int, v: int, twoway: bool):
        self.map[u].append(v)
        if twoway:
            self.map[v].append(u)


class CharUnweightedGrid(UnweightedGraph):

    deltas = ((0, -1), (0, 1), (-1, 0), (1, 0))

    # s: list[str] と書くと、AtCoder で RE になります。
    def __init__(self, s: list, wall='#'):
        self.h = len(s)
        self.w = len(s[0])
        super().__init__(self.h * self.w)
        self.s = s
        self.wall = wall

    def to_vertex_id(self, i: int, j: int):
        return self.w * i + j

    def find_vertex_id(self, c: str):
        for i in range(self.h):
            for j in range(self.w):
                if self.s[i][j] == c:
                    return self.w * i + j
        return -1

    def get_edges(self, v: int):
        i, j = v // self.w, v % self.w
        for di, dj in self.deltas:
            ni, nj = i + di, j + dj
            if 0 <= ni and ni < self.h and 0 <= nj and nj < self.w and self.s[ni][nj] != self.wall:
                yield self.w * ni + nj
